File: browser/src/js/moneysocket/protocol/transact/provider_nexus.js
// Copyright (c) 2020 Jarret Dyrbye
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php

const Timestamp = require('../../utl/timestamp.js').Timestamp;
const ProtocolNexus = require("../nexus.js").ProtocolNexus;

const NotifyInvoice = require(
    "../../message/notification/invoice.js").NotifyInvoice;
const NotifyPreimage = require(
    "../../message/notification/preimage.js").NotifyPreimage;


const LAYER_REQUESTS = new Set(["REQUEST_PAY",
                                "REQUEST_INVOICE",
                               ]);

class ProviderTransactNexus extends ProtocolNexus {
    constructor(below_nexus, layer) {
        super(below_nexus, layer);
        console.assert(typeof this.layer.requestInvoiceCb == 'function');
        console.assert(typeof this.layer.requestPayCb == 'function');
    }

    ///////////////////////////////////////////////////////////////////////////

    handleLayerRequest(msg) {
        if (msg['request_name'] == "REQUEST_INVOICE") {
            this.layer.requestInvoiceCb(this, msg['msats'],
                                        msg['request_uuid']);
        } else if (msg['request_name'] == "REQUEST_PAY") {
            this.layer.requestPayCb(this, msg['bolt11'], msg['request_uuid']);
        }
    }

    isLayerMessage(msg) {
        if (msg['message_class'] != "REQUEST") {
            return false;
        }
        return LAYER_REQUESTS.has(msg['request_name']);
    }

    recvFromBelowCb(below_nexus, msg) {
        //console.log("provider transact nexus got message");
        if (! this.isLayerMessage(msg)) {
            super.recvFromBelowCb(below_nexus, msg)
            return;
        }
        this.handleLayerRequest(msg);
    }

    recvRawFromBelowCb(below_nexus, msg_bytes) {
        //console.log("transact nexus got raw msg from below");
    }

    notifyInvoice(bolt11, request_reference_uuid) {
        this.send(new NotifyInvoice(bolt11, request_reference_uuid));
    }

    notifyPreimage(preimage, request_reference_uuid) {
        this.send(new NotifyPreimage(preimage, null, request_reference_uuid));
    }

}

exports.ProviderTransactNexus = ProviderTransactNexus;

File: python/moneysocket/protocol/transact/provider_nexus.py
# Copyright (c) 2020 Jarret Dyrbye
# Distributed under the MIT software license, see the accompanying
# file LICENSE or http://www.opensource.org/licenses/mit-license.php

from moneysocket.protocol.nexus import ProtocolNexus
from moneysocket.message.notification.invoice import NotifyInvoice
from moneysocket.message.notification.preimage import NotifyPreimage


LAYER_REQUESTS = {"REQUEST_PAY", "REQUEST_INVOICE"}

class ProviderTransactNexus(ProtocolNexus):
    def __init__(self, below_nexus, layer):
        super().__init__(below_nexus, layer)
        assert "request_invoice_cb" in dir(self.layer)
        assert "request_pay_cb" in dir(self.layer)

    def handle_layer_request(self, msg):
        if msg['request_name'] == "REQUEST_INVOICE":
            self.layer.request_invoice_cb(self, msg['msats'],
                                          msg['request_uuid'])
        elif msg['request_name'] == "REQUEST_PAY":
            self.layer.request_pay_cb(self, msg['bolt11'], msg['request_uuid'])

    def is_layer_message(self, msg):
        if msg['message_class'] != "REQUEST":
            return False
        return msg['request_name'] in LAYER_REQUESTS

    def recv_from_below_cb(self, below_nexus, msg):
        if not self.is_layer_message(msg):
            super().recv_from_below_cb(below_nexus, msg)
            return
        self.handle_layer_request(msg)

    def recv_raw_from_below_cb(self, below_nexus, msg_bytes):
        pass

    def notify_invoice(self, bolt11, request_reference_uuid):
        self.send(NotifyInvoice(bolt11, request_reference_uuid))

    def notify_preimage(self, preimage, request_reference_uuid):
        self.send(NotifyPreimage(preimage, None, request_reference_uuid))
